etcdserver/v3_server.go

```diff
@@ -641,3 +646,89 @@ func (s *EtcdServer) processInternalRaftRequest(ctx context.Context, r pb.Intern
 
 // Watchable returns a watchable interface attached to the etcdserver.
 func (s *EtcdServer) Watchable() mvcc.WatchableKV { return s.KV() }
+
+func (s *EtcdServer) linearizableReadLoop() {
+	var (
+		rs       raft.ReadState
+		requests int
+	)
+	ctx := make([]byte, 8)
+	s.readNotifyc = make(chan struct{})
+
+	for {
+		binary.BigEndian.PutUint64(ctx, s.reqIDGen.Next())
+
+		for (len(s.readwaitc) > 0 && requests < 256) || (len(s.readwaitc) == 0 && requests == 0) {
```

Comment:
I think this can all go away with `s.readwaitc = make(chan struct{}, 1)`?
```go
// wait for initial waiter on current s.readyNotifyc
select {
case <-s.readwaitc:
case <-s.done: return
}
s.readMu.Lock()
notifyc := s.readNotifyc
s.readNotifyc = make(chan struct{})
// any remaining waiters at this point are waiting on notifyc
select {
case <-s.readwaitc:
default:
}
s.readMu.Unlock()
```